src/LiveComponent/assets/src/live_controller.ts

```diff
@@ -243,8 +251,9 @@ export default class extends Controller {
      * @param {boolean} shouldRender Whether a re-render should be triggered
      * @param {string|null} extraModelName Another model name that this might go by in a parent component.
      * @param {Object} options Options include: {bool} dispatch
+     * @param {any} modelValue Original HTML element value (for handling collection checkbox fields)
      */
-    $updateModel(model: string, value: any, shouldRender = true, extraModelName: string | null = null, options: any = {}) {
+    $updateModel(model: string, value: any, shouldRender = true, extraModelName: string | null = null, options: any = {}, modelValue: any = null) {
```

Comment:
Your logic is all sound - I was able to play with your version of the demo and it all works perfectly (I'd like to incorporate the newsletter into the field after this is merged).

But, I think we need to reorganize the code a little bit. Let's look and think about JUST this method. It's called `$updateModel`. And so, it should be that simple: we pass it the model name and value, and it takes care of the rest. And so, if you think about calling this method programmatically, if we were to update the `newsletter` field, it should look like this:

```
liveController.$updateModel('registration_form[newsletter][]', ['mailing']);
```

Notice I pass in the model name - `registration_form[newsletter][]` - and then the FINAL value - `['mailing']`.

So what I'm saying is this:

A) Remove the `modelValue` argument from this method.
B) Instead, in `_updateModelFromElement`, we need to handle the logic of calculating the correct value from there. For example, if we determine that I just unchecked `notification`, we would grab (all in `_updateModelFromElement()`) the current value for the `registration_form[newsletter][]` model and *remove* `notification`. Basically, your logic from `set_deep_data.ts` would move to `_updateModelFromElement()`. This will give us the final value, and we can call `$updateModel` like normal.
C) All the changes in `set_deep_data.ts` can be removed, because they will be done in `_updateModelFromElement()`.

Let me know if that makes sense - we can also chat on Symfony's Slack.

Reply:
Btw, we'll also need a test for the checkboxes - basically you'll want to replicate what you did in your live demo in our test suite here. 

The proper place for the test is `model.test.ts`. You should create my own "template" inside the test itself like is done in `standardizes user[firstName] style models into post.name` with the two checkboxes. You can then "check" one of those boxes, and assert (via that `mockRerender()` function) that the correct query params are sent. You might even create a couple of tests, like starting with both checked, and unchecking one. Let me know if you need some guidance - those tests can be weird at first.

Reply:
It all makes perfect sense. I will probably come back with refactored version (hopefully with tests) next week and I'll make sure to join Sf Slack in case I need someone to explain to me what happens in those tests 👍

Reply:
I have done refactoring part. As you know my tests (at least JS parts) are not functioning as expected yet, but hopefully with some help I'll get there.

During refactor I found it useful to extract traversing deepData out of `setDeepData` so that I could easily access data without code duplication. In addition I had to make a small change in `normalizeModelName` as trailing `[]` should be stripped during normalization unconditionally IMO - in HTML forms `[]` at the end of the name is only used to mark form elements that should be sent as an array.

Other changes are I think in line with what you suggested.